source/source.go

```diff
@@ -53,35 +59,64 @@ func (s *Source) Configure(ctx context.Context, cfg map[string]string) error {
 	return err
 }
 
+// Open prepare the plugin to start sending records from the given position
 func (s *Source) Open(ctx context.Context, pos sdk.Position) error {
 	logger := sdk.Logger(ctx).With().Str("Class", "Source").Str("Method", "Open").Logger()
 	logger.Trace().Msg("Starting Open the Source Connector...")
+	p, err := position.ParseRecordPosition(pos)
+	if err != nil {
+		logger.Error().Stack().Err(err).Msg("Error While parsing the Position")
+		return err
+	}
+	logger.Info().Msgf("Requested fields: %s", s.config.Fields)
 
 	config := minimarketo.ClientConfig{
 		ID:       s.config.ClientID,
 		Secret:   s.config.ClientSecret,
-		Endpoint: s.config.Endpoint,
-		Debug:    true,
-		// Debug: false,
+		Endpoint: s.config.ClientEndpoint,
+		Debug:    false,
 	}
-	var err error
-	s.client, err = minimarketo.NewClient(config)
+	s.client, err = marketoclient.NewClient(config)
 	if err != nil {
-		logger.Error().Stack().Err(err).Msg("Error While Creating the Client")
+		logger.Error().Stack().Err(err).Msg("Error While Creating the Marketo Client")
 		return err
 	}
+	s.iterator, err = iterator.NewCombinedIterator(ctx, s.config.ClientEndpoint, s.config.PollingPeriod, s.client, p, s.config.Fields)
+	if err != nil {
+		logger.Error().Stack().Err(err).Msg("Error while create a combined iterator")
+		return fmt.Errorf("couldn't create a combined iterator: %w", err)
+	}
 	logger.Trace().Msg("Successfully Created the Source Connector")
 	return nil
 }
 
+// Read gets the next record from the Marketo Instance
 func (s *Source) Read(ctx context.Context) (sdk.Record, error) {
-	return s.iterator.Next(ctx)
+	logger := sdk.Logger(ctx).With().Str("Class", "Source").Str("Method", "Read").Logger()
+	logger.Trace().Msg("Starting Read the Source Connector...")
+
+	if !s.iterator.HasNext(ctx) {
+		logger.Debug().Msg("No more records to read, sending sdk.ErrorBackoff...")
+		return sdk.Record{}, sdk.ErrBackoffRetry
+	}
+
+	record, err := s.iterator.Next(ctx)
+	if err != nil {
+		logger.Error().Stack().Err(err).Msg("Error while fetching the records")
+		return sdk.Record{}, err
+	}
+	return record, nil
 }
 
 func (s *Source) Ack(ctx context.Context, pos sdk.Position) error {
+	sdk.Logger(ctx).Debug().Str("position", string(pos)).Msg("got ack")
 	return nil
 }
 
 func (s *Source) TearDown(ctx context.Context) error {
+	if s.iterator != nil {
+		s.iterator.Stop()
+		s.iterator = nil
```

Comment:
If there's a special reason for setting the iterator to nil, we should mention it.

Reply:
updated.